README.md
# casa-on-rails-cookbook

offers an `lwrp` for setting up a casa-on-rails instance.

## Supported Platforms

CentOS 6

## Attributes

## Usage

Instantiate a casa on rails instance with this resource.  Note the defaults are commented. 

```ruby
casa_on_rails 'default' do
  # repo 'https://github.com/ucla/casa-on-rails.git'
  # revision '1.1.12'
  # port 3000
  # run_user 'casa' # if != 'casa', will not be added to run_group on creation.
  # run_group 'casa'
  # db_host '127.0.0.1'
  # db_port 3306
  # db_name 'casa'
  # db_user 'casa'
  # db_password 'tsktsk'
  # es_host '127.0.0.1'
  # es_port 9200
  # es_index 'casa'
  deploy_path '/var/casa'
  # bundler_path nil
  # rails_env 'production'
  uuid  'de6edba2-c398-44bd-8438-9df2aa70a5b7'
  secret 'cookiesecretusesomethingrandom'
  contact_name 'Joe Schmoe'
  contact_email 'joe@schmoe.edu'
  # action :create
end
```

You can start/stop the resulting rails app using `service casa-default` (the name of the resource is used in the service name)

## Assumptions
Just like the casa-on-rails app, it's still early days.  I've made a few assumptions here to simplify the process:

  * You will resolve `mysql` and `elasticsearch` dependencies yourself. Take a look in `fixtures/cookbooks/casa_on_rails_test` for some hints on how I'd do that.
  * You will pass the path to your `bundle` bin file. `sysvinit` on centos is extremely minimal in it's path, so if you don't give me your path, the service is probably not going to start! :). Additionally, I have no interest in controlling how you deploy ruby, so casa_on_rails wont configure a ruby for itself!

## License and Authors

License:: Apache 2.0, see `LICENSE` file for text.
Author:: Steve Nolen (<technolengy@gmail.com>)


CHANGELOG.md
# 0.4.0

Support running casa instances as different users. Change license to apache 2.0.

# 0.1.0

Initial release of casa-on-rails


test/fixtures/cookbooks/casa_on_rails_test/metadata.rb
name 'casa_on_rails_test'
version '0.0.1'

depends 'casa-on-rails'
depends 'mysql', '~> 6.0'
depends 'ruby_rbenv', '~> 1.0.1'
depends 'ruby_build'
depends 'nodejs'


test/fixtures/cookbooks/casa_on_rails_test/recipes/delete.rb
# sets up casa on rails before running this integration which tears it back down.
include_recipe 'casa_on_rail_test::default'

casa_on_rails 'default' do
  deploy_path '/var/casa'
  secret '0d7e46be6a8fd3a1e4cc3b11d8a09a03d7c948f3dc772119cce2'
  uuid '988ab9d1-6e7c-44b1-9273-65d378941f54'
  contact_name 'Steve Nolen'
  contact_email 'technolengy@gmail.com'
  bundler_path '/usr/local/rbenv/shims'
  action :delete
end

test/fixtures/cookbooks/casa_on_rails_test/recipes/default.rb
# install ruby with rbenv, java, npm, git, mysql-server and set up db.
node.default['rbenv']['rubies'] = ['2.1.7']
include_recipe 'ruby_build'
include_recipe 'ruby_rbenv::system'
include_recipe 'nodejs::npm'
package 'java-1.7.0-openjdk'
package 'git'
rbenv_global '2.1.7'
rbenv_gem 'bundle'

mysql_service 'default' do
  port '3306'
  version '5.6'
  initial_root_password 'changeme'
  action [:create, :start]
end

execute 'add test db info' do
  command "sleep 5s; /usr/bin/mysql -h 127.0.0.1 -uroot -pchangeme -e \"CREATE DATABASE IF NOT EXISTS casa; GRANT ALL ON casa.* to 'casa' identified by 'tsktsk';\""
end

# actual casa service block
casa_on_rails 'default' do
  deploy_path '/var/casa'
  secret '0d7e46be6a8fd3a1e4cc3b11d8a09a03d7c948f3dc772119cce2'
  uuid '988ab9d1-6e7c-44b1-9273-65d378941f54'
  contact_name 'Steve Nolen'
  contact_email 'technolengy@gmail.com'
  bundler_path '/usr/local/rbenv/shims'
  es_host 'elasticsearch' # linked docker container
end


libraries/provider_casa_on_rails.rb
require 'chef/provider/lwrp_base'
require_relative 'helpers'

class Chef
  class Provider
    class CasaOnRails < Chef::Provider::LWRPBase # rubocop:disable ClassLength
      # Chef 11 LWRP DSL Methods
      use_inline_resources if defined?(use_inline_resources)

      def whyrun_supported?
        true
      end

      # Mix in helpers from libraries/helpers.rb
      include CasaOnRailsCookbook::Helpers

      action :create do
        # casa user
        group "#{new_resource.name} :create casa" do
          group_name new_resource.run_group
          action :create
        end

        user "#{new_resource.name} :create casa" do
          username new_resource.run_user
          gid 'casa' if new_resource.run_user == 'casa'
          action :create
        end

        # init file for service, abstract to support deb and rhel7
        template "/etc/init.d/casa-#{new_resource.name}" do
          owner 'root'
          group 'root'
          mode '0755'
          source 'sysvinit.erb'
          cookbook 'casa-on-rails'
          variables(
            config: new_resource,
            name: new_resource.name,
            app_path: "#{new_resource.deploy_path}/current",
            port: new_resource.port,
            rails_env: new_resource.rails_env,
            ruby_exec_path: new_resource.bundler_path
          )
        end

        # add shared dirs for chef deploy
        %w(config pids log).each do |d|
          directory "#{new_resource.deploy_path}/shared/#{d}" do
            recursive true
            owner new_resource.run_user
            group new_resource.run_group
          end
        end

        # database.yml
        template "#{new_resource.deploy_path}/shared/config/database.yml" do
          source 'database.yml.erb'
          owner new_resource.run_user
          group new_resource.run_group
          cookbook 'casa-on-rails'
          variables(
            casa_db_password: new_resource.db_password,
            casa_db_user: new_resource.db_user,
            casa_db_name: new_resource.db_name,
            casa_db_host: new_resource.db_host,
            casa_db_port: new_resource.db_port
          )
          notifies :restart, "service[casa-#{new_resource.name}]", :delayed
        end

        # secrets
        template "#{new_resource.deploy_path}/shared/config/secrets.yml" do
          source 'secrets.yml.erb'
          owner new_resource.run_user
          group new_resource.run_group
          cookbook 'casa-on-rails'
          variables(
            casa_secret: new_resource.secret
          )
          notifies :restart, "service[casa-#{new_resource.name}]", :delayed
        end

        # generate casa config.
        template "#{new_resource.deploy_path}/shared/config/casa.yml" do
          source 'casa.yml.erb'
          owner new_resource.run_user
          group new_resource.run_group
          cookbook 'casa-on-rails'
          variables(
            casa_uuid: new_resource.uuid,
            casa_contact_name: new_resource.contact_name,
            casa_contact_email: new_resource.contact_email
          )
          notifies :restart, "service[casa-#{new_resource.name}]", :delayed
        end

        # generate ES config file, only supports one instance currently.
        template "#{new_resource.deploy_path}/shared/config/elasticsearch.yml" do
          source 'elasticsearch.yml.erb'
          owner new_resource.run_user
          group new_resource.run_group
          cookbook 'casa-on-rails'
          variables(
            casa_es_host: new_resource.es_host,
            casa_es_port: new_resource.es_port,
            casa_es_index: new_resource.es_index
          )
          notifies :restart, "service[casa-#{new_resource.name}]", :delayed
        end

        # required headers for mysql2 gem (which gets installed with bundler below)
        package 'mysql-devel'

        # farm out to chef deploy. TODO: make an attribute that only does this once.
        # note namespace "new resource" causes some weird stuff here.
        computed_path = path_plus_bundler
        casa_resource = new_resource
        deploy_branch casa_resource.name do
          deploy_to casa_resource.deploy_path
          repo casa_resource.repo
          revision casa_resource.revision
          user casa_resource.run_user
          group casa_resource.run_group
          symlink_before_migrate(
            'config/database.yml' => 'config/database.yml',
            'config/casa.yml' => 'config/casa.yml',
            'config/elasticsearch.yml' => 'config/elasticsearch.yml',
            'config/secrets.yml' => 'config/secrets.yml',
            'bundle' => '.bundle'
          )
          before_migrate do
            execute 'bundle install' do
              environment 'PATH' => computed_path
              cwd release_path
              command "bundle install --path #{casa_resource.deploy_path}/shared/bundle"
            end
            execute 'npm install' do
              cwd release_path
            end
            execute 'block build' do
              environment 'PATH' => computed_path
              cwd release_path
              command 'bundle exec blocks build'
            end
          end
          migrate true
          migration_command "RAILS_ENV=#{casa_resource.rails_env} bundle exec rake db:migrate"
          purge_before_symlink %w(log tmp/pids public/system config/database.yml config/secrets.yml)
          before_symlink do
            execute 'db:seed' do
              environment 'PATH' => computed_path
              cwd release_path
              command "RAILS_ENV=#{casa_resource.rails_env} bundle exec rake db:seed; touch #{casa_resource.deploy_path}/shared/.seeded"
              not_if { ::File.exist?("#{casa_resource.deploy_path}/shared/.seeded") }
            end
          end
          restart_command "service casa-#{casa_resource.name} restart"
        end

        service "casa-#{new_resource.name}" do
          supports restart: true, status: true
          action [:enable, :start]
        end
      end

      action :delete do
        # stop service
        service "casa-#{new_resource.name}" do
          supports restart: true, status: true
          action [:disable, :stop]
        end
        
        # delete deploy path and remove init script.
        directory "#{new_resource.deploy_path}" do
          action :delete
        end
        
        file "/etc/init.d/casa-#{new_resource.name}" do
          action :delete
        end
      end
    end
  end
end


libraries/resource_casa_on_rails.rb
require 'chef/resource/lwrp_base'

class Chef
  class Resource
    class CasaOnRails < Chef::Resource::LWRPBase
      self.resource_name = :casa_on_rails
      actions :create, :delete
      default_action :create

      attribute :name, kind_of: String, name_attribute: true
      attribute :repo, kind_of: String, default: 'https://github.com/ucla/casa-on-rails.git'
      attribute :revision, kind_of: String, default: '114d4a86dc67567abbba9fbff8366f175551e99d'
      attribute :port, kind_of: Integer, default: 3000
      attribute :run_user, kind_of: String, default: 'casa'
      attribute :run_group, kind_of: String, default: 'casa'
      attribute :db_host, kind_of: String, default: '127.0.0.1'
      attribute :db_port, kind_of: Integer, default: 3306
      attribute :db_name, kind_of: String, default: 'casa' # set to name attr?
      attribute :db_user, kind_of: String, default: 'casa'
      attribute :db_password, kind_of: String, default: 'tsktsk'
      attribute :es_host, kind_of: String, default: '127.0.0.1'
      attribute :es_port, kind_of: Integer, default: 9200
      attribute :es_index, kind_of: String, default: 'casa'
      attribute :deploy_path, kind_of: String, required: true
      attribute :bundler_path, kind_of: String, default: nil
      attribute :rails_env, kind_of: String, default: 'production'
      attribute :uuid, kind_of: String, required: true, regex: [/[a-f0-9]{8}-[a-f0-9]{4}-4[a-f0-9]{3}-[89aAbB][a-f0-9]{3}-[a-f0-9]{12}/i]
      attribute :secret, kind_of: String, required: true
      attribute :contact_name, kind_of: String, required: true
      attribute :contact_email, kind_of: String, required: true
    end
  end
end
